data/src/main/java/com/linkedin/data/codec/entitystream/AbstractDataDecoder.java

```diff
@@ -0,0 +1,354 @@
+          case BOOL_TRUE:
+            addValue(Boolean.TRUE);
+            break;
+          case BOOL_FALSE:
+            addValue(Boolean.FALSE);
+            break;
+          case NULL:
+            addValue(Data.NULL);
+            break;
+          case NOT_AVAILABLE:
+            readNextChunk();
+            return;
+          default:
+            handleException(new Exception("Unexpected token " + token + " from data parser"));
+        }
+      }
+    }
+    catch (IOException e)
+    {
+      handleException(e);
+    }
+  }
+
+  /**
+   * Interface to new complex object, invoked at the start of parsing object.
+   */
+  protected abstract DataComplex createDataObject(NonBlockingDataParser parser);
+
+  /**
+   * Interface to new complex list, invoked at start of parsing array.
+   */
+  protected abstract DataComplex createDataList(NonBlockingDataParser parser);
+
+  protected final boolean isCurrList()
+  {
+    return _isCurrList;
+  }
+
+  protected final void validate(NonBlockingDataParser.Token token)
+  {
+    if (!(token == NOT_AVAILABLE || _expectedTokens.contains(token)))
+    {
+      handleException(new Exception("Expecting " + _expectedTokens + " but got " + token));
+    }
+  }
+
+  private void push(DataComplex dataComplex, boolean isList)
+  {
+    if (!(_isCurrList || _stack.isEmpty()))
+    {
+      _currFieldStack.push(_currField);
+      _currField = null;
+    }
+    _stack.push(dataComplex);
+    _isCurrList = isList;
+    updateExpected();
+  }
+
+  @SuppressWarnings("unchecked")
+  private void pop()
+  {
+    // The stack should never be empty because of token validation.
+    assert !_stack.isEmpty() : "Trying to pop empty stack";
+
+    DataComplex tmp = _stack.pop();
+    tmp = postProcessDataObject(tmp);
+    if (_stack.isEmpty())
+    {
+      _result = (T) tmp;
+      // No more tokens is expected.
+      _expectedTokens = NONE;
+    }
+    else
+    {
+      _isCurrList = _stack.peek() instanceof DataList;
+      if (!_isCurrList)
+      {
+        _currField = _currFieldStack.pop();
+      }
+      addValue(tmp);
+      updateExpected();
+    }
+  }
+
+  protected DataComplex postProcessDataObject(DataComplex dataComplex)
+  {
+    return dataComplex;
+  }
+
+  protected void addValue(Object value)
+  {
+    if (!_stack.isEmpty())
+    {
+      DataComplex currItem = _stack.peek();
+      if (_isCurrList)
+      {
+        CheckedUtil.addWithoutChecking((DataList) currItem, value);
+      }
+      else
+      {
+        addEntryToDataObject(currItem, _currField, value);
+        _currField = null;
+      }
+      updateExpected();
+    }
+  }
+
+  protected void addEntryToDataObject(DataComplex dataObject, String currField, Object currValue)
+  {
+    CheckedUtil.putWithoutChecking((DataMap) dataObject, currField, currValue);
+  }
+
+  protected final void updateObjectStackPeek(DataComplex dataComplex)
+  {
+    _stack.pop();
+    _stack.push(dataComplex);
+  }
+  
+  protected void updateExpected()
```

Comment:
This can be private? If not add doc

Reply:
added doc, cannot make this private as this was already exposed